portalnet/src/utp_controller.rs

```diff
@@ -15,33 +22,235 @@ use crate::discovery::UtpEnr;
 pub struct UtpController {
     pub inbound_utp_transfer_semaphore: Arc<Semaphore>,
     pub outbound_utp_transfer_semaphore: Arc<Semaphore>,
-    /// uTP socket.
     utp_socket: Arc<UtpSocket<UtpEnr>>,
+    metrics: OverlayMetricsReporter,
+}
+
+lazy_static! {
+    /// The default configuration to use for uTP connections.
+    pub static ref UTP_CONN_CFG: ConnectionConfig = ConnectionConfig { max_packet_size: 1024, ..Default::default()};
 }
 
 impl UtpController {
-    pub fn new(utp_transfer_limit: usize, utp_socket: Arc<UtpSocket<UtpEnr>>) -> Self {
+    pub fn new(
+        utp_transfer_limit: usize,
+        utp_socket: Arc<UtpSocket<UtpEnr>>,
+        metrics: OverlayMetricsReporter,
+    ) -> Self {
         Self {
             utp_socket,
             inbound_utp_transfer_semaphore: Arc::new(Semaphore::new(utp_transfer_limit)),
             outbound_utp_transfer_semaphore: Arc::new(Semaphore::new(utp_transfer_limit)),
+            metrics,
+        }
+    }
+
+    /// Connect with a peer given the connection id, and return the data received from the peer.
+    pub async fn connect_inbound_stream(
+        &self,
+        connection_id: u16,
+        peer: Enr,
+    ) -> anyhow::Result<Vec<u8>> {
+        self.metrics
+            .report_utp_active_inc(UtpDirectionLabel::Inbound);
+        let cid = utp_rs::cid::ConnectionId {
+            recv: connection_id,
+            send: connection_id.wrapping_add(1),
+            peer: UtpEnr(peer),
+        };
+        let stream = match self.connect_with_cid(cid.clone(), *UTP_CONN_CFG).await {
```

Comment:
This is my personal preference. If you don't like, feel free to ignore it.

I don't like the use of `match` for `Option` or `Result` when one case is trivial. I find the following code more readable.

```rust
let stream = self.connect_with_cid(...)
    .await
    .map_err(|err| {
        ...
        anyhow!(...)
    })?;
```

This applies to many places in this file.

Reply:
Yup, I'd agree with that style guideline, updated the code. Except for in places where we need to return a non-error value from within the closure (eg. a bool inside `outbound_stream`), which I don't think is possible to return early from inside a closure in Rust.